akka-cluster-typed/src/test/scala/akka/cluster/ddata/typed/scaladsl/ReplicatorSpec.scala

```diff
@@ -34,59 +34,47 @@ object ReplicatorSpec {
-  private case class InternalChanged[A <: ReplicatedData](chg: Replicator.Changed[A]) extends InternalMsg
 
   val Key = GCounterKey("counter")
 
   def client(replicator: ActorRef[Replicator.Command])(implicit cluster: Cluster): Behavior[ClientCommand] =
     Actor.deferred[ClientCommand] { ctx ⇒
-      val updateResponseAdapter: ActorRef[Replicator.UpdateResponse[GCounter]] =
-        ctx.spawnAdapter(InternalUpdateResponse.apply)
-
-      val getResponseAdapter: ActorRef[Replicator.GetResponse[GCounter]] =
-        ctx.spawnAdapter(InternalGetResponse.apply)
-
-      val changedAdapter: ActorRef[Replicator.Changed[GCounter]] =
-        ctx.spawnAdapter(InternalChanged.apply)
-
-      replicator ! Replicator.Subscribe(Key, changedAdapter)
+      replicator ! Replicator.Subscribe(Key, ctx.self.upcast)
 
       def behavior(cachedValue: Int): Behavior[ClientCommand] = {
         Actor.immutable[ClientCommand] { (ctx, msg) ⇒
           msg match {
             case Increment ⇒
-              replicator ! Replicator.Update(Key, GCounter.empty, Replicator.WriteLocal, updateResponseAdapter)(_ + 1)
+              replicator ! Replicator.Update(Key, GCounter.empty, Replicator.WriteLocal, ctx.self.upcast)(_ + 1)
               Actor.same
 
             case GetValue(replyTo) ⇒
-              replicator ! Replicator.Get(Key, Replicator.ReadLocal, getResponseAdapter, Some(replyTo))
+              replicator ! Replicator.Get(Key, Replicator.ReadLocal, ctx.self.upcast, Some(replyTo))
               Actor.same
 
             case GetCachedValue(replyTo) ⇒
-              replicator ! Replicator.Get(Key, Replicator.ReadLocal, getResponseAdapter, Some(replyTo))
+              replicator ! Replicator.Get(Key, Replicator.ReadLocal, ctx.self.upcast, Some(replyTo))
               Actor.same
-
-            case internal: InternalMsg ⇒ internal match {
-              case InternalUpdateResponse(_) ⇒ Actor.same // ok
-
-              case InternalGetResponse(rsp @ Replicator.GetSuccess(Key, Some(replyTo: ActorRef[Int] @unchecked))) ⇒
-                val value = rsp.get(Key).value.toInt
-                replyTo ! value
-                Actor.same
-
-              case InternalGetResponse(rsp) ⇒
-                Actor.unhandled // not dealing with failures
-
-              case InternalChanged(chg @ Replicator.Changed(Key)) ⇒
-                val value = chg.get(Key).value.intValue
-                behavior(value)
-            }
           }
         }
       }
+        .onResponse[Replicator.GetResponse[GCounter]] { (_, msg) ⇒
+          msg match {
+            case rsp @ Replicator.GetSuccess(Key, Some(replyTo: ActorRef[Int] @unchecked)) ⇒
+              val value = rsp.get(Key).value.toInt
+              replyTo ! value
+              Actor.same
+            case _ ⇒
+              Actor.unhandled // not dealing with failures
+          }
+        }
+        .onResponse[Replicator.Changed[GCounter]] { (ctx, chg) ⇒
```

Comment:
i guess these will only be used for responses but there's now nothing stopping someone from casting an actor ref or using a type that is compatible with the actor type and using this for other mischievous things

Reply:
that's true, lot's of nasty things can be done by a programmer ;-)
at least `ref.asInstanceOf[SomeOtherType]` smells pretty bad